polynote-server/src/main/scala/polynote/server/repository/fs/LocalFilesystem.scala

```diff
@@ -22,9 +23,55 @@ class LocalFilesystem(maxDepth: Int = 4) extends NotebookFilesystem {
     content <- readBytes(is).ensuring(ZIO.effectTotal(is.close()))
   } yield new String(content.toArray, StandardCharsets.UTF_8)
 
+  /**
+    * Writes the content to the given path "atomically" – at least, in a way where the existing file at that path (if
+    * present) won't be unrecoverable if writing fails for some reason.
+    */
+  private def writeAndSwap(path: Path, content: Array[Byte]) =
+    try {
+      // first, try creating the file in case it doesn't exist. If it doesn't, go ahead and write directly to it.
+      // it's not great to use an exception for normal flow control, but I don't think there's another way to avoid a
+      // race condition here (e.g. if we checked that the file doesn't exist, but then it does by the time we go to
+      // write to it)
+      Files.createFile(path)
+      Files.write(path, content)
+    } catch {
+      case _: FileAlreadyExistsException =>
+        // the file already exists, so we're going to write to a temporary file next to it and then swap it.
+        // this is to prevent unrecoverable incomplete writes if the application dies during the write.
+        val filename = path.getFileName.toString.stripPrefix(".")
+        val tmpPath = path.resolveSibling(s".$filename.swp")
+        Files.write(tmpPath, content)
+
+        // try an atomic move first to replace the old file with the temp file
+        try {
+          Files.move(tmpPath, path, StandardCopyOption.ATOMIC_MOVE, StandardCopyOption.REPLACE_EXISTING)
+        } catch {
+          case err: AtomicMoveNotSupportedException =>
+            // couldn't move atomically; instead copy the original file to an "old" path, move the new file to the path,
+            // then delete the old. Then, if the new file doesn't write successfully, then the original is there at the
+            // "old" path. AFAIK this code path shouldn't ever happen on a POSIX-compliant system.
+            val oldPath = path.resolveSibling(s".$filename.old")
+            Files.copy(path, oldPath, StandardCopyOption.REPLACE_EXISTING)
+
+            try {
+              Files.move(tmpPath, path)
+            } catch {
+              case err: Throwable =>
+                // failed to move the new file to the path; try to copy the old file back
+                Files.copy(oldPath, path)
+
+                // rethrow; don't try to delete the old file in case something is weird.
+                throw new IOException(s"Failed to move new file over $path; original file is copied at $oldPath in case of an issue.", err)
+            }
+            Files.delete(oldPath)
```

Comment:
is it possible for the swp file to remain if the move commands fail? should we try to delete it just in case?

Reply:
In the case where things start going wrong here, I'd rather leave as many files behind as possible so the user can conceivably do something to recover the one they want?